.github/workflows/main.yml

```diff
@@ -126,56 +113,129 @@ jobs:
         run: yarn
       - working-directory: ./web
         run: npm test
-      - uses: codecov/codecov-action@v1
+      - uses: codecov/codecov-action@a1ed4b322b4b38cb846afb5a0ebfa17086917d27
+        # mirrored above and at https://github.com/mitmproxy/mitmproxy/settings/actions
         with:
           file: ./web/coverage/coverage-final.json
           name: web
 
   docs:
+    if: github.repository != 'mitmproxy/mitmproxy' || github.ref != 'refs/heads/main'
+    # mirrored below >>>
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          persist-credentials: false
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.9'
+      - run: python release/pins/pip-install-pinned.py
+      - run: |
+          wget -q https://github.com/gohugoio/hugo/releases/download/v0.83.1/hugo_extended_0.83.1_Linux-64bit.deb
+          echo "9487ea3b80f8ddd0ba600d42850b96b6a8b0bb9b41bc08cb285635ebbd41328d hugo_extended_0.83.1_Linux-64bit.deb" | sha256sum -c
+          sudo dpkg -i hugo*.deb
+      - run: ./build.py
+        working-directory: docs
+    # <<<
+
+# All actions below this line produce stuff that may eventually be deployed,
+# so we make sure to really pin everything here.
+
+  deploy-docs:
+    if: github.repository == 'mitmproxy/mitmproxy' && github.ref == 'refs/heads/main'
     environment: deploy-docs
     env:
       AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
       AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+    # mirrored above >>>
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v1
+      - uses: actions/checkout@v2
+        with:
+          persist-credentials: false
       - uses: actions/setup-python@v2
         with:
           python-version: '3.9'
-      - run: pip install tox
+      - run: python release/pins/pip-install-pinned.py
       - run: |
-          wget https://github.com/gohugoio/hugo/releases/download/v0.70.0/hugo_extended_0.70.0_Linux-64bit.deb
+          wget -q https://github.com/gohugoio/hugo/releases/download/v0.83.1/hugo_extended_0.83.1_Linux-64bit.deb
+          echo "9487ea3b80f8ddd0ba600d42850b96b6a8b0bb9b41bc08cb285635ebbd41328d hugo_extended_0.83.1_Linux-64bit.deb" | sha256sum -c
           sudo dpkg -i hugo*.deb
-      - run: tox -e docs
+      - run: ./build.py
+        working-directory: docs
+    # <<<
+      - run: ./upload-dev.sh
+        working-directory: docs
+
+  build:
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - image: macos-10.15
+            platform: macos
+          - image: windows-2019
+            platform: windows
+          - image: ubuntu-16.04  # Old Ubuntu version for old glibc
```

Comment:
16.04 LTS is now EOL since a few weeks.
When was the last time we got a user complaining about glibc issues? 
I'd be happy to see this bumped to 18.04 or even 20.04.